### PR TITLE
Rust: Implement `UnusedVariable.ql`

### PR DESCRIPTION
**Pull request by hvitved**
Since this query is based entirely on syntax, we can implement it already now without SSA.

### CODE REVIEW

File: rust/ql/lib/codeql/rust/elements/internal/VariableImpl.qll [hunk 1]
```diff
@@ -98,6 +98,27 @@ module Impl {
...
+    /**
+     * Gets the pattern that declares this variable.
+     *
+     * Normally, the pattern is unique, except when introduced in an or pattern:
+     *
+     * ```rust
+     * match either {
+     *    Either::Left(x) | Either::Right(x) => println!(x),
+     * }
+     * ```
```
Comment (geoffw0) [inline rust/ql/lib/codeql/rust/elements/internal/VariableImpl.qll:111]: So in that case we would get two results for `getPat`, namely the two `x`s on the left-hand side of the `=>`?

File: rust/ql/lib/codeql/rust/elements/internal/VariableImpl.qll [hunk 2]
```diff
@@ -366,6 +387,29 @@ module Impl {
     override string getAPrimaryQlClass() { result = "VariableAccess" }
   }
 
+  /** Holds if `e` occurs in the LHS of a (compound) assignment. */
+  private predicate assignLhs(Expr e) {
+    exists(BinaryExpr be |
+      be.getOperatorName().regexpMatch(".*=") and
```
Comment (geoffw0) [inline rust/ql/lib/codeql/rust/elements/internal/VariableImpl.qll:393]: I don't feel strongly about this one, but I believe `matches` is still preferred over `regexpMatch` where possible???
```suggestion
      be.getOperatorName().matches("%=") and
```

File: rust/ql/lib/codeql/rust/elements/internal/VariableImpl.qll [hunk 3]
```diff
@@ -366,6 +387,29 @@ module Impl {
...
+      e = be.getLhs()
+    )
+    or
+    exists(Expr mid |
+      assignLhs(mid) and
+      getImmediateParent(e) = mid
+    )
```
Comment (geoffw0) [inline rust/ql/lib/codeql/rust/elements/internal/VariableImpl.qll:400]: What situation is this case intended to cover?

File: rust/ql/lib/codeql/rust/elements/internal/VariableImpl.qll [hunk 4]
```diff
@@ -366,6 +387,29 @@ module Impl {
     override string getAPrimaryQlClass() { result = "VariableAccess" }
   }
 
+  /** Holds if `e` occurs in the LHS of a (compound) assignment. */
```
Comment (geoffw0) [inline rust/ql/lib/codeql/rust/elements/internal/VariableImpl.qll:390]: I think we mean a compound or non-compound assignment, but the brackets aren't communicating that to me clearly.
```suggestion
  /** Holds if `e` occurs in the LHS of an assignment or compound assignment. */
```

File: rust/ql/lib/codeql/rust/elements/internal/VariableImpl.qll [hunk 5]
```diff
@@ -366,6 +387,29 @@ module Impl {
...
+
+  /** A variable write. */
+  class VariableWriteAccess extends VariableAccess {
+    VariableWriteAccess() { assignLhs(this) }
+  }
+
+  /** A variable read. */
+  class VariableReadAccess extends VariableAccess {
+    VariableReadAccess() { not this instanceof VariableWriteAccess }
+  }
```
Comment (geoffw0) [inline rust/ql/lib/codeql/rust/elements/internal/VariableImpl.qll:411]: Isn't a write in a compound assignment also a read?

File: rust/ql/test/query-tests/unusedentities/main.rs [hunk 6]
```diff
@@ -1,118 +1,119 @@
...
+    let js = [1, 2, 3];
+    let ks = [1, 2, 3];
 
-	println!("lets use {:?}", js);
+    println!("lets use {:?}", js);
 
-	for k in ks {
-		println!("lets use {}", k);
-	}
+    for k // BAD: unused variable [SPURIOUS: macros not yet supported]
```
Comment (geoffw0) [inline rust/ql/test/query-tests/unusedentities/main.rs:112]: This one shouldn't be marked as `BAD`.
```suggestion
    for k // SPURIOUS: unused variable (macros not yet supported)
```

**Review by geoffw0 — COMMENTED** [review]
LGTM.

A few questions and optional suggestions.

File: rust/ql/lib/codeql/rust/elements/internal/VariableImpl.qll (hunk 1 above)
Comment (hvitved) [reply rust/ql/lib/codeql/rust/elements/internal/VariableImpl.qll:111]: Yes.

File: rust/ql/lib/codeql/rust/elements/internal/VariableImpl.qll (hunk 3 above)
Comment (hvitved) [reply rust/ql/lib/codeql/rust/elements/internal/VariableImpl.qll:400]: Destructuring assignments: https://doc.rust-lang.org/reference/expressions/operator-expr.html#destructuring-assignments

File: rust/ql/lib/codeql/rust/elements/internal/VariableImpl.qll (hunk 5 above)
Comment (hvitved) [reply rust/ql/lib/codeql/rust/elements/internal/VariableImpl.qll:411]: Yes

File: rust/ql/lib/codeql/rust/elements/internal/VariableImpl.qll (hunk 2 above)
Comment (geoffw0) [reply rust/ql/lib/codeql/rust/elements/internal/VariableImpl.qll]: Yep, there was a discussion a long while back about whether the optimiser is more aware of `matches`, I can't remember the answer.  Anyway, up to you.

File: rust/ql/lib/codeql/rust/elements/internal/VariableImpl.qll (hunk 4 above)
Comment (geoffw0) [reply rust/ql/lib/codeql/rust/elements/internal/VariableImpl.qll]: (addressed)

File: rust/ql/lib/codeql/rust/elements/internal/VariableImpl.qll (hunk 5 above)
Comment (geoffw0) [reply rust/ql/lib/codeql/rust/elements/internal/VariableImpl.qll:420]: (addressed)

Thanks.

**Review by geoffw0 — APPROVED** [review]
LGTM.

I believe this is our first working problem query!  ✨

**Pull request merged**